docs: update the blob module specs part 1 (#1906)
celestiaorg/celestia-app: x/blob/README.md

## Abstract

The `x/blob` module enables users to pay for arbitrary data to be published to
the Celestia blockchain. Users create a single `BlobTx` that is composed of:

1. Multiple `Blob`s (Binary Large OBjects): the data they wish to publish. A
   single `Blob` is composed of:
    1. `NamespaceId  []byte`: the namespace this blob should be published to.
    1. `Data         []byte`: the data to be published.

>     1. `ShareVersion uint32`: the version of the share format used to encode

Reviewer: I don't think we have one yet although I definitely agree that anyplace we mention some versioning we should be clear about what constitutes a version change and especially for this specific parameter because it's user facing

>     1. `ShareVersion uint32`: the version of the share format used to encode

Reviewer: I think this is the proto format which only does uint32. We should specify that these values are capped to 256 values (i.e. uint8)

>     1. `ShareVersion uint32`: the version of the share format used to encode

Author: added a link to ADR014, and mentioned that we specify there how it effects the share encoding and when that value would get incrememnted

       this blob into a share.

> 1. A single `sdk.Tx` which is composed of:

Reviewer: Can we first split this into:
```
1. A single `sdk.Tx` which is composed of:
   1. One and only one `MsgPayForBlob`
   2. Any other sdk messages.
```

And then in the PFB part explain the subcomponents

Reviewer: Ah no wait ignore, I realise we don't support multiple other sdk messages in a `BlobTx`

> 1. A single `sdk.Tx` which is composed of:

Author: added a link to the sdk ADR that discusses the structure of sdk.Tx, but perhaps we can talk about this more elsewhere? I'm quite sure what to do yet.

    1. `Signer string`: the transaction signer
    1. `NamespaceIds []byte`: the namespaces they wish to publish each blob to.
       The namespaces here must match the namespaces in the `Blob`s.
    1. `ShareCommitment []byte`: a share commitment that is the root of a Merkle
       tree where the leaves are share commitments to each blob associated with

>        this BlobTx.

Author: 1e600c166c8e30ecb1ffcfc5caa22e3d1e19f293

nice catch

After the `BlobTx` is submitted to the network, a block producer separates the
transaction from the blob. Both components get included in the data square in
different namespaces: the BlobTx gets included in the PayForBlobNamespace (one
of the [reserved
namespaces](../../specs/src/specs/consensus.md#reserved-namespaces)) and the
associated blob gets included in the namespace the user specified in the
original `BlobTx`. Further reading: [Message Block
Layout](https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md)

After a block has been created, the user can verify that their data was included

> in a block via a blob inclusion proof. A blob inclusion proof uses the

Author: we don't have one yet, but hopefully soon!

> in a block via a blob inclusion proof. A blob inclusion proof uses the

Reviewer 2: I see, then may I suggest adding an empty link just as a placeholder, and later filling it with the proper link? or a todo as a comment?

> in a block via a blob inclusion proof. A blob inclusion proof uses the

Author: added a todo in https://github.com/celestiaorg/celestia-app/pull/1906/commits/5cc0296392c05e48716544da5b66b1aa86358636 and specifically mentioned this comment in #1962

`ShareCommitment` in the original transaction and subtree roots of the block's
data square to prove to the user that the shares that compose their original
data do in fact exist in a particular block.

## State

> The blob module doesn't maintain it's own state outside of a single param.

Author: went ahead and just added a params section, and tried to further specify that we only use the paramstore and auth stores

When a `MsgPayForBlob` is processed, it consumes gas based on the blob size.

## Messages

- [`MsgPayForBlobs`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/proto/celestia/blob/v1/tx.proto#L16-L31)

>   pays for a set of blobs to be included in the block. Transactions containing

Reviewer 2: Thanks for the explanation, then considering this, I think the following sentence in the specs needs some revision:
```suggestion
  pays for a set of blobs to be included in the block. Blob transactions that contain
  this `sdk.Msg` are also referred to as "PFBs".
```

  this `sdk.Msg` are better known as "PFBs".

> ```proto

Reviewer 2: This protobuf definition is a great addition 👍

// MsgPayForBlobs pays for the inclusion of a blob in the block.
message MsgPayForBlobs {
  string signer = 1;
  // namespaces is a list of namespaces that the blobs are associated with. A
  // namespace is a byte slice of length 33 where the first byte is the

>   // namespaceVersion and the subsequent 32 bytes are the namespaceId.

Reviewer 3: oof noticed that `33` is incorrect. Fixed in https://github.com/celestiaorg/celestia-app/pull/1951

  repeated bytes namespaces = 2;

>   repeated uint32 blob_sizes = 3;

Reviewer 2: Would be good to clarify whether sizes are reported in bits or bytes.

  // share_commitments is a list of share commitments (one per blob).
  repeated bytes share_commitments = 4;
  // share_versions are the versions of the share format that the blobs
  // associated with this message should use when included in a block. The
  // share_versions specified must match the share_versions used to generate the

>   // share_commitment in this message.

Author: added a link to adr014

> ```proto
> // MsgPayForBlobs pays for the inclusion of a blob in the block.
> message MsgPayForBlobs {
>   string signer = 1;
>   // namespaces is a list of namespaces that the blobs are associated with. A
>   // namespace is a byte slice of length 33 where the first byte is the
>   // namespaceVersion and the subsequent 32 bytes are the namespaceId.
>   repeated bytes namespaces = 2;
>   repeated uint32 blob_sizes = 3;
>   // share_commitments is a list of share commitments (one per blob).
>   repeated bytes share_commitments = 4;
>   // share_versions are the versions of the share format that the blobs
>   // associated with this message should use when included in a block. The
>   // share_versions specified must match the share_versions used to generate the
>   // share_commitment in this message.
>   repeated uint32 share_versions = 8;
> }
> ```

Reviewer 2: I would also like to propose an alternative format (completely optional) that can potentially improve the organization of the text.  We can keep the proto definition of the fields as it is, but separate the comments for each field and provide an explanation for each field in the text i.e.,

```suggestion

```proto
message MsgPayForBlobs {
  string signer = 1;
  repeated bytes namespaces = 2;
  repeated uint32 blob_sizes = 3;
  repeated bytes share_commitments = 4;
  repeated uint32 share_versions = 8;
}

MsgPayForBlobs pays for the inclusion of a blob in the block and consists of the following fields:
- signer: ...
- namespace: namespace is a byte slice of length 33 where the first byte is the
  namespaceVersion and the subsequent 32 bytes are the namespaceId.
- blob_sizes: ...
- share_commitments is a list of share commitments (one per blob).
- share_versions are the versions of the share format that the blobs associated with this message should use when included in a block. The share_versions specified must match the share_versions used to generate the share_commitment in this message.
```

### Generating the `ShareCommitment`

> 1. Split the blob into shares of size `appconsts.ShareSize`

Reviewer 2: A bit of context here would be helpful, for example, who generates the share commitments? what is the purpose of generating share commitments? etc.

Author: https://github.com/celestiaorg/celestia-app/pull/1906/commits/145e3ade71560b6b47284c68360b900e2b9d6860

> 1. Split the blob into shares of size `appconsts.ShareSize`

Reviewer 2: A link to the `appconsts.ShareSize` would be great. 
Also, if this constant is  covered in other specs, a link to that specs would be even better (instead of/alongside  `appconsts.ShareSize` where  `appconsts` refers to a go package).

1. Determine the
   [`SubtreeWidth`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/pkg/shares/non_interactive_defaults.go#L94-L116)
   by dividing the length in shares by the `SubtreeRootThreshold`.
1. Generate each subtree root by diving the blob shares into `SubtreeWidth`
   sized sets, then take the binary [namespaced merkle tree
   (NMT)](https://github.com/celestiaorg/nmt/blob/v0.16.0/docs/spec/nmt.md) root
   of each set of shares.
1. Calculate the final share commitment by taking the merkle root (note: not an
   NMT, just a normal binary merkle root) of the subtree roots from the previous
   step.

See
[`CreateCommitment`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/x/blob/types/payforblob.go#L169-L236)
for an implementation. See [data square
layout](../../specs/src/specs/data_square_layout.md) and
[ADR013](../../docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md)
for details on the rational of the square layout.

## Validity Rules

In order for a proposal block to be considered valid, each `BlobTx`, and thus
each PFB, to be included in a block must follow a set of validity rules.

> 1. Signatures: All blob transactions must have valid signatures. This is

Reviewer 2: The signature field has not been explained yet, or did I miss it?

Reviewer: I guess this more relates to the SDK then something specific to the blob module

   state-dependent because correct signatures require using the correct sequence

>    number(aka nonce).

Reviewer 2: Would be helpful to include a link to where signature generation and nonces are explained.

> 1. Single SDK.Msg: There must be only a single sdk.Msg in the blob transaction.

Reviewer 2: ```suggestion
1. Single SDK.Msg: There must be only a single sdk.Msg encoded in the `sdk.Tx` field of the blob transaction `BlobTx`.
```

> 1. Namespace Validity: The namespace of each blob transaction must be valid.

Reviewer 2: ```suggestion
1. Namespace Validity: The namespace of each blob in a blob transaction `BlobTx` must be valid.
```

>    This validity is determined by the following sub-rules:

Reviewer 2: For accuracy's sake, it is good to specify the denote the formal field name that each subrule applies to.

>     1. The namespace is not within the reserved namespace range.

Reviewer 2: The namespace appears in two places one in the `MsgPayForBlobs.NamespaceIds` field and one in the `Blob.NamespaceId`, would be good to clarify that this rule applies to both of them.

Author: https://github.com/celestiaorg/celestia-app/pull/1906/commits/60e4cce0113c93c386a03ce3be97a7aee6a36062

tried to add some clarity by adding another rule that is enforced. The namespaces in the blob and the PFB have to be the same, and if they are the same, then they will both be invalid or valid

>     1. The namespace is not the tail padding or parity namespaces.

Reviewer 2: Would be great to add links for each different reserved namespace mentioned in these two lines.

1. Blob Size: No blob can have a size of 0.
1. Blob Count: There must be one or more blobs included in the transaction.
1. Share Commitment Validity: Each share commitment must be valid.
    1. The size of each of the share commitments must be equal to the digest of
       the hash function used (sha256 so 32 bytes).
    1. The share commitment must be calculated using the steps specified above
       in [Generating the Share
       Commitment](./README.md#generating-the-sharecommitment)

> 1. Share Versions: The versions of the shares must be supported.

Reviewer 2: A link to where supported share versions are explained would be nice.

Author: good, idea, added a link to ADR014 and mentioned that it contains info on what it does to the encoding and when it gets incremented

1. Signer Address: The signer address must be a valid Celestia address.
1. Proper Encoding: The blob transactions must be properly encoded.
1. Size Consistency: The sizes included in the transaction must match the actual

>    sizes of the blobs.

Reviewer 2: A bit of elaboration on this would be helpful for accuracy, e.g., when we say, is it the byte size or the bit size? also, the exact name of the field of the `BlobTx` to which this rule applies.

Author: :+1: https://github.com/celestiaorg/celestia-app/pull/1906/commits/4338e36444725cb94055ff3c406181e29a541571

## Abstract

The `x/blob` module enables users to pay for arbitrary data to be published to
the Celestia blockchain. Users create a single `BlobTx` that is composed of:

1. Multiple `Blob`s (Binary Large OBjects): the data they wish to publish. A
   single `Blob` is composed of:
    1. `NamespaceId  []byte`: the namespace this blob should be published to.
    1. `Data         []byte`: the data to be published.
    1. `ShareVersion uint32`: the version of the share format used to encode
       this blob into a share.

> 1. A single [`sdk.Tx`](https://github.com/celestiaorg/cosmos-sdk/blob/v1.15.0-sdk-v0.46.13/docs/architecture/adr-020-protobuf-transaction-encoding.md) which is composed of:

Reviewer 2: [nit] 
```suggestion
1. A single [`sdk.Tx`](https://github.com/celestiaorg/cosmos-sdk/blob/v1.15.0-sdk-v0.46.13/docs/architecture/adr-020-protobuf-transaction-encoding.md) which encapsulates a `MsgPayFroBlob` message that is composed of:
```

Reviewer 3: [nit^2]

typo in `MsgPayFroBlob` also I think there's only a plural `MsgPayForBlobs`

Reviewer 2: Thanks, fixed it :)

    1. `ShareCommitments [][]byte`: a share commitment that is the root of a Merkle
       tree where the leaves are share commitments to each blob associated with
       this `BlobTx`.

After the `BlobTx` is submitted to the network, a block producer separates the
transaction i.e., `sdk.Tx` from the blob. Both components get included in the data square in
different namespaces: the `sdk.Tx` of the original `BlobTx` together with some metadata about the separated blobs get included in the PayForBlobNamespace (one
of the [reserved
namespaces](../../specs/src/specs/consensus.md#reserved-namespaces)) and the
associated blob gets included in the namespace the user specified in the
original `BlobTx`. Further reading: [Data Square Layout](../../specs/src/specs/data_square_layout.md)

After a block has been created, the user can verify that their data was included
in a block via a blob inclusion proof. A blob inclusion proof uses the
`ShareCommitment` in the original `sdk.Tx` transaction and subtree roots of the block's
data square to prove to the user that the shares that compose their original
data do in fact exist in a particular block.

## State

> The blob module doesn't maintain it's own state outside of a two params. Meaning

Reviewer 3: ```suggestion
The blob module doesn't maintain it's own state outside of two params. Meaning
```

that the blob module only uses the params and auth module stores.

### Params

```proto
// Params defines the parameters for the module.
message Params {
  option (gogoproto.goproto_stringer) = false;
  uint32 gas_per_blob_byte = 1
      [ (gogoproto.moretags) = "yaml:\"gas_per_blob_byte\"" ];
  uint64 gov_max_square_size = 2
      [ (gogoproto.moretags) = "yaml:\"gov_max_square_size\"" ];
}
```

#### `GasPerBlobByte`

`GasPerBlobByte` is the amount of gas that is consumed per byte of blob data

> when a `MsgPayForBlob` is processed. Currently, the Default value is 8. This

Reviewer 3: ```suggestion
when a `MsgPayForBlobs` is processed. Currently, the default value is 8. This
```

value is set below that of normal transaction gas consumption, which is 10.

#### `GovMaxSquareSize`

`GovMaxSquareSize` is the maximum size of a data square that is considered valid
by the validator set. This value is superseded by the `MaxSquareSize`, which is
hardcoded and cannot change without hardforking the chain. See
[ADR021](../../docs/architecture/adr-021-restricted-block-size.md) for more
details.

## Messages

- [`MsgPayForBlobs`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/proto/celestia/blob/v1/tx.proto#L16-L31)
  pays for a set of blobs to be included in the block. Blob transactions that contain
  this `sdk.Msg` are also referred to as "PFBs".

>   this `sdk.Msg` are better known as "PFBs".

Reviewer 3: ```suggestion
```

```proto
message MsgPayForBlobs {
  string signer = 1;
  repeated bytes namespaces = 2;
  repeated uint32 blob_sizes = 3;
  repeated bytes share_commitments = 4;
  repeated uint32 share_versions = 8;
}
```

> `MsgPayForBlobs` pays for the inclusion of a blob in the block and consists of the

Reviewer 2: [nit] 
```suggestion
`MsgPayForBlobs` pays for the inclusion of blobs in the block and consists of the
```

following fields:

- signer: bech32 encoded signer address

> - namespace: namespace is a byte slice of length 33 where the first byte is the
>   namespaceVersion and the subsequent 32 bytes are the namespaceId.

Reviewer 3: [blocking]

```suggestion
- namespace: namespace is a byte slice of length 29 where the first byte is the
  namespaceVersion and the subsequent 28 bytes are the namespaceId.
```

Author: ahh shoot, I purposfully didn't update since I saw https://github.com/celestiaorg/celestia-app/pull/1951/files :facepalm: but forgot that this was copied pasted. thanks for catching

- blob_sizes: sizes of each blob in bytes.
- share_commitments is a list of share commitments (one per blob).
- share_versions are the versions of the share format that the blobs associated
  with this message should use when included in a block. The share_versions
  specified must match the share_versions used to generate the share_commitment
  in this message. See
  [ADR014](../../docs/architecture/adr-014-versioned-namespaces.md) for more

>   details on how this effects the share encoding and when it is updated.

Reviewer 3: [optional] ADR-014 doesn't describe share versions, it describe namespace versions.

ADR-007 introduces a share version: https://github.com/celestiaorg/celestia-app/blob/main/docs/architecture/adr-007-universal-share-prefix.md

Author: my bad thanks for catching https://github.com/celestiaorg/celestia-app/pull/1906/commits/920760cb312985deed1e99c8a6f681a8a4c24f29

1. Determine the
   [`SubtreeWidth`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/pkg/shares/non_interactive_defaults.go#L94-L116)
   by dividing the length in shares by the `SubtreeRootThreshold`.
1. Generate each subtree root by diving the blob shares into `SubtreeWidth`
   sized sets, then take the binary [namespaced merkle tree
   (NMT)](https://github.com/celestiaorg/nmt/blob/v0.16.0/docs/spec/nmt.md) root
   of each set of shares.
1. Calculate the final share commitment by taking the merkle root (note: not an
   NMT, just a normal binary merkle root) of the subtree roots from the previous
   step.

See
[`CreateCommitment`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/x/blob/types/payforblob.go#L169-L236)
for an implementation. See [data square
layout](../../specs/src/specs/data_square_layout.md) and
[ADR013](../../docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md)
for details on the rational of the square layout.

## Validity Rules

In order for a proposal block to be considered valid, each `BlobTx`, and thus
each PFB, to be included in a block must follow a set of validity rules.

> 1. Signatures: All blob transactions must have valid signatures. This is

Reviewer 2: [Question for my own understanding] When the term "proposal block" is used in the context of "In order for a proposal block to be considered valid," does it refer to a block proposed during the consensus (and yet to be voted and committed) or one that has received the majority vote (hence committed)? I'm curious about when signatures and nonces are verified – during the voting phase or at the time of block execution (which I think corresponds to the processProposal). If signature verification is required during the voting process, then we would need to track the state changes resulting from proposals (but not committed blocks) in order to do so.

Author: > does it refer to a block proposed during the consensus (and yet to be voted and committed) or one that has received the majority vote (hence committed)?

both! we call ProcessProposal on proposals beofore voting, and while consensus nodes sync the chain

signatures, and therefore their nonces, are verified before the block is actually applied to the state in both cases. State can then be tracked rather easily since we already have the state from the previous block, which is all that is needed.
